### PR TITLE
Spanner repository factory

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
Added spanner repository factory and tests

### CODE REVIEW

**Comment by ChengyuanZhao** [comment]
@balopat PTAL, these are the moved-over boilerplate files from the other PR :D

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryBeanTests.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.data.repository.core.support.RepositoryFactorySupport;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+
+/**
+ * @author Chengyuan Zhao
+ */
+@RunWith(SpringRunner.class)
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryBeanTests.java]: this is not used, you can remove it

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Chengyuan Zhao
+ */
+@RunWith(SpringRunner.class)
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java]: this is not used, you can remove it

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		SpannerOperations operations = mock(SpannerOperations.class);
+		new SpannerRepositoryImpl(operations, Object.class).count();
+		verify(operations, times(1)).count(eq(Object.class));
+	}
+
+	@Test
+	public void deleteByIdTest() {
+		SpannerOperations operations = mock(SpannerOperations.class);
+		String key = "key";
+		Object ret = new Object();
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java]: nit: unused variable, please remove

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	@Test
+	public void getTargetRepositoryTest() {
+		RepositoryInformation repoInfo = mock(RepositoryInformation.class);
+		when(repoInfo.getRepositoryBaseClass())
+				.thenReturn((Class) SpannerRepositoryImpl.class);
+		when(repoInfo.getDomainType()).thenReturn((Class) TestEntity.class);
+		Object repo = this.spannerRepositoryFactory.getTargetRepository(repoInfo);
+		assertTrue(SpannerRepository.class.isAssignableFrom(repo.getClass()));
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java]: I'm not sure why you need to use `isAssignableFrom`? It seems that we should expect the exact class returned from the repoInfo

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Chengyuan Zhao
+ */
+@RunWith(SpringRunner.class)
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java]: not required.

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	@Test
+	public void getSpannerOperationsTest() {
+		SpannerOperations operations = mock(SpannerOperations.class);
+		assertSame(operations, new SpannerRepositoryImpl(operations, Object.class)
+				.getSpannerOperations());
+	}
+
+	@Test(expected = IllegalArgumentException.class)
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java]: this is great! I think we are missing the same test checks for thrown `IllegalArgumentException` for null argument for the others:
````
	void deleteById(ID id);
	void delete(T entity);
	void deleteAll(Iterable<? extends T> entities);
	<S extends T> Iterable<S> saveAll(Iterable<S> entities);
	Optional<T> findById(ID id);
	boolean existsById(ID id);
````

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+		Object ob = new Object();
+		new SpannerRepositoryImpl(operations, Object.class).delete(ob);
+		verify(operations, times(1)).delete(eq(ob));
+	}
+
+	@Test
+	public void deleteManyObsTest() {
+		SpannerOperations operations = mock(SpannerOperations.class);
+		Iterable<String> obs = Arrays.asList("ob1", "ob2");
+		doAnswer(new Answer() {
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java]: replace with lambda?

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java (hunk 4 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java]: I think all we need to know is that it's returning a SpannerRepository , not any specific implementation.

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java (hunk 7 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryImplTests.java]: @balopat changes all made from this pass!

File: spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java (hunk 4 above)
Comment (balopat) [reply spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/SpannerRepositoryFactoryTests.java]: We'll need the actual equality on `SpannerRepositoryImpl.class`, based on the instructions in the [RepositoryFactorySupport javadoc](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/core/support/RepositoryFactorySupport.html#getRepositoryBaseClass-org.springframework.data.repository.core.RepositoryMetadata-)

````
	/**
	 * Returns the base class backing the actual repository instance. Make sure
	 * {@link #getTargetRepository(RepositoryInformation)} returns an instance of this class.
	 *
	 * @param metadata
	 * @return
	 */
	protected abstract Class<?> getRepositoryBaseClass(RepositoryMetadata metadata);
````

**Comment by balopat** [comment]
LGTM
